### DISCUSSION TITLE
Azure VM Tasks (non-ACI)

### DISCUSSION THREAD

**Discussion by n1hility** [cirruslabs/cirrus-ci-docs#953] [Ideas]
Not sure if this should be a separate issue than #778. I have a use case that requires nested virt on Windows. GCP's nested virt only supports Linux. EC2 only supports it with cost-prohibitive bare metal nodes. Azure on the other hand supports nested virt on the Dv3 and Ev3 nodes.  It would be great to have a direct vm option, similar to ec2-instance that works with azure to make usage of these nodes.

**Comment by fkorotkov** [comment]
Seems similar to #778. @n1hility, I'm curious what is your use case for nested virtualization on Windows?

**Comment by n1hility** [comment]
@fkorotkov Sure glad to provide details. I am contributing a patch to the podman project that enables windows support (https://github.com/containers/podman/pull/12503). The implementation is based off of WSL, which internally uses hyper-v to run the linux kernel in a special shared VM. We need the ability to test this implementation as part of pull request checking/CI. AFAICT the only approach that works with Cirrus CI atm is a persistent worker. Alternatively, I suppose it could be possible to doubly nest inside of a google compute L1 linux node with windows running under KVM. Although that introduces some challenges with upkeep on the windows instance.

**Reply by fkorotkov** [reply]
Thank you for the details. I'll take a look into this early next week after the Christmas long weekend.

**Reply by n1hility** [reply]
@fkorotkov thanks for taking a look, I hope you have a wonderful holiday with your family

**Reply by n1hility** [reply]
hi @fkorotkov just following up on this one. Just curious if you had a chance to look at this and if you had if you planned to add it to your roadmap. Thanks!

**Reply by fkorotkov** [reply]
Sorry for the delay. It felt off my radar. Thank you for pinging again!

I've investigate the issue and seems there are some hick-ups. Seems there is [no way to create such VM with a custom startup script in one API request](https://github.com/Azure/azure-sdk-for-java/issues/1910#issuecomment-1024982379) like we for GCP, AWS and Oracle Cloud. I need to either figure if there is a way to provide a custom script or we'll need to change how Cirrus interacts with such services. Right now Cirrus has an assumption that to schedule a task only one API request is needed.

**Reply by n1hility** [reply]
Hey @fkorotkov, I just wanted to follow up on this one.  I realize this ended up being more involved than hoped, but I just wanted to see if you had decided to add it to your future plans. Thanks again!

**Reply by fkorotkov** [reply]
Hey @n1hility, sorry, but at the moment we don't have bandwidth. Preparing something up for WWDC. We can take another look at it later in June after the WWDC.

Not ideal, but as a workaround you can try to configure a [persistent worker](https://cirrus-ci.org/guide/persistent-workers/).

**Reply by n1hility** [reply]
Hey @fkorotkov, I hope WWDC went well! I totally understand you guys have priorities and this is just one of the many things going on. I hope you don't mind me checking in again (and from time to time). I am just trying to get an idea for our own planning. Thanks so much!

**Reply by fkorotkov** [reply]
Hey @n1hility, we are a little overwhelmed with other things like bringing all GCP specific feature for AWS as well to have a feature parity between the clouds. On that note I was wondering if [AWS' Metal option](https://aws.amazon.com/blogs/compute/running-hyper-v-on-amazon-ec2-bare-metal-instances/) will work for you. It seems you are already getting some AWS Graviton instances to your fleet in https://github.com/containers/aardvark-dns.

**Reply by n1hility** [reply]
@fkorotkov Ok thanks for the update. Unfortunately, metal instances are exceptionally cost-prohibitive. The cheapest Windows metal instance is around 60 times the cost of a standard azure instance (6-7$ an hour vs 10cents for a Dv3 node typical for CI), so not really practical for pull request verification. EC2 Arm metal nodes aren't licensed for/support Windows, Although even if so, the metal nodes are all Graviton v2 / ARMv8.2, so missing FEAT_NV* needed for nested virt.

**Reply by fkorotkov** [reply]
Got it. Azure instances that is. Once #979 and #900 are closed I promise Azure stuff will be next.

**Reply by n1hility** [reply]
Awesome! Thanks @fkorotkov!

**Reply by n1hility** [reply]
hey @fkorotkov, just wondering if this one is still on your roadmap, and if so when you think it might land. Thanks so much!

**Comment by evertonlperes** [comment]
Hi there,
Just jumping in this thread. 
Also, the kvm enabled on Windows runners will be very appreciated. I do have the almost the same scenario as @n1hility. If you need further details, I'll be happy to share @fkorotkov. Thanks!
